From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nesaak <52047222+Nesaak@users.noreply.github.com>
Date: Tue, 12 May 2020 17:00:24 -0400
Subject: [PATCH] Add PersistentDataContainer getter in ItemStack

Adds a new method which gets the PersistentDataContainer directly from an ItemStack.
This method is an alternative for cloning ItemMeta, getting the PDC and setting the new ItemMeta which can be much slower.

getPersistentDataContainer - Gets the ItemStacks PersistentDataContainer. If the ItemStack has no meta, it creates a new meta for it.


diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
index d7d6a3e83dfd88359708749f5c12be02815c3580..6c988104c5ee63bb0604dcdcaae2d4cd5fb7fcec 100644
--- a/src/main/java/org/bukkit/inventory/ItemStack.java
+++ b/src/main/java/org/bukkit/inventory/ItemStack.java
@@ -15,6 +15,7 @@ import org.bukkit.enchantments.Enchantment;
 import org.bukkit.inventory.meta.Damageable;
 import org.bukkit.inventory.meta.ItemMeta;
 import org.bukkit.material.MaterialData;
+import org.bukkit.persistence.PersistentDataContainer; // Paper
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
@@ -767,5 +768,15 @@ public class ItemStack implements Cloneable, ConfigurationSerializable {
         ItemMeta itemMeta = getItemMeta();
         return itemMeta.hasItemFlag(flag);
     }
+
+    /**
+     * Gets the {@link PersistentDataContainer} of this items current {@link ItemMeta}. Can be faster than cloning ItemMeta using {@link #getItemMeta()}.
+     *
+     * @return The current PeristentDataContainer of this item's meta
+     */
+    @NotNull
+    public PersistentDataContainer getPersistentDataContainer() {
+        return (meta == null ? meta = Bukkit.getItemFactory().getItemMeta(type) : meta).getPersistentDataContainer();
+    }
     // Paper end
 }
